src/lib/AuthManager.js

```diff
@@ -0,0 +1,236 @@
+        // _and_ nextcloud has stopped sending new cookies, stop sending login data
+        this.needsAuthentication = false
+      }
+    } else {
+      this.everRecievedCookie = true
+    }
+  }
+}
+
+// Handles listeners for AuthSessions. Should have exactly one global instance.  
+export class AuthManager {
+
+  constructor() {
+    this.sessions = {} // dict: AuthSessionId => AuthSession
+    this.requests = {} // dict: request.requestId => AuthSessionId
+    this.sessionIdState = 0
+
+    this.onResponseListener = e => this.onResponse(e)
+    this.onRequestListener = e => this.onRequest(e)
+  }
+
+  updateListeners() {
+    let filter = this.getRequestFilter()
+    if (filter.urls.length <= 0) {
+      this.removeListeners()
+    } else {
+      this.addListeners(filter)
+    }
+  }
+
+  // adds or overwrites listeners
+  addListeners(filter) {
+
+    let respPerms = ["blocking", "responseHeaders"]
+    let reqPerms = ["blocking", "requestHeaders"]
+
+    // "extraHeaders" is required additionally for chrome to show relevant cookies
+    if (browser.isChrome) {
+      respPerms.push("extraHeaders")
+      reqPerms.push("extraHeaders")
+    }
+
+    browser.webRequest.onHeadersReceived.addListener(
+      this.onResponseListener,
+      filter,
+      respPerms
+    )
+
+    browser.webRequest.onBeforeSendHeaders.addListener(
+      this.onRequestListener,
+      filter,
+      reqPerms
+    )
+  }
+
+  removeListeners() {
+    browser.webRequest.onHeadersReceived.removeListener(this.onResponseListener)
+    browser.webRequest.onBeforeSendHeaders.removeListener(this.onRequestListener)
+  }
+
+  onResponse(e) {
+      let header = e.responseHeaders
+      // get authSession by request id
+      let authSessionId = this.requests[e.requestId]
+      if (authSessionId === undefined) {
+        // request wasn't handled by this
+        console.info("Detected a response not meant for the bookmarks API. Returning now.")
+      } else {
+        // save cookie in session
+        this.sessions[authSessionId].acceptCookies(header)
+        // remove request and its ownership
+        delete this.requests[e.requestId]
+        delete this.sessions[authSessionId].ownedRequests[e.requestId]
+      }
+      return {responseHeaders: e.responseHeaders};
+  }
+
+  onRequest(e) {
+      let header = e.requestHeaders
+      // get the authSessionId
+      let authSessionId
+      for (let object of header) {
+        if (object.name.toLowerCase() == "authsessionid") {
+          authSessionId = object.value
+        }
+      }
+
+      if ( authSessionId === undefined ) {
+        // not a request managed by this
+        console.info("Detected a request not meant for the bookmarks API. Returning now.")
+      } else {
+        // assign the requestId a authSession
+        this.requests[e.requestId] = authSessionId
+        // set ownership of request
+        this.sessions[authSessionId].ownedRequests[e.requestId] = true
```

Comment:
Is this tracking still necessary?

Reply:
Every request adds an entry `request.id: authSessionId` into `authManager.requests` [line 152](https://github.com/pogobanane/floccus/blob/cookie-hack/src/lib/AuthManager.js#L153)

If a request get a response, the request.id entry is removed. [line 132](https://github.com/pogobanane/floccus/blob/cookie-hack/src/lib/AuthManager.js#L132)

If a request doesn't get a response, the entry has to be removed when the authSession is destroyed. Therefore to be able to remove entries on destruction, the authSession has to know, which `request.id`s it owns. 

As AuthManager is a long-lived object, we should make sure it doesn't memory leak.

Reply:
Aaah, it's used over here: https://github.com/marcelklehr/floccus/pull/494/files#diff-646ada40e3b3d2b59ae3f458ee99355eR125 sorry for the noise :D